File: recursion/permutation/java/Permutation.java
import java.util.*;

public class Permutation {
    public static List<int[]> perms(int[] A) {
	List<int[]> soln = new ArrayList<int[]>();
	perms(A, 0, soln);
	return soln;
    }

    private static void perms(int[] A, int pos, List<int[]> soln) {
	if (pos == A.length) {
	    int[] cur = new int[A.length];
	    System.arraycopy(A, 0, cur, 0, A.length);
	    soln.add(cur);
	    return;
	}
	for (int i = pos; i < A.length; i++) {
	    swap(A, pos, i);
	    perms(A, pos+1, soln);
	    swap(A, pos, i);
	}
    }

    private static void swap(int[] A, int i, int j) {
	int t = A[i];
	A[i] = A[j];
	A[j] = t;
    }
}

File: recursion/permutation/permutation.cpp
#include "permutation.h"

void PermutationRec(std::vector<int>& elems,
                    int i,
                    std::vector<std::vector<int>>& perms) {
  int n = elems.size();
  if (i == n) {
    // Hit the base case.
    perms.emplace_back(elems.begin(), elems.end());
    return;
  }
  for (int j = i; j < n; ++j) {
    std::swap(elems[i], elems[j]);
    PermutationRec(elems, i+1, perms);
    std::swap(elems[i], elems[j]);
  }
}

// Permutation takes an array of elems, and output all its
// permutations in *perms*.
void Permutation(const std::vector<int>& elems,
                 std::vector<std::vector<int>>& perms) {
  if (elems.empty()) {
    return;
  }
  std::vector<int> mutable_elems(elems);
  PermutationRec(mutable_elems, 0, perms);
}
